### PR TITLE
When a given function is deserialized multiple times in the same process, ensure the same object is returned.

### PR DESCRIPTION
**Pull request by pitrou**
This avoids repetitive compilation of the same signatures.

### CODE REVIEW

File: numba/dispatcher.py [hunk 1]
```diff
@@ -433,23 +438,49 @@ def __reduce__(self):
...
         self._can_compile = can_compile
         return self
 
+    @property
+    def _uuid(self):
+        """
+        An instance-specific UUID, to avoid multiple deserializations of
+        a given instance.
+
+        Note this is lazily-generated, for performance reasons.
```
Comment (jcrist) [inline numba/dispatcher.py]: Is the performance hit of generating a uuid on `Dispatcher` creation really that bad (actually curious, not sure what kind of overhead you expect for `Dispatcher` creation alone)? On my computer:

```python
In [1]: from uuid import uuid1

In [2]: %timeit uuid1()
100000 loops, best of 3: 6.29 µs per loop
```

**Comment by jcrist** [comment]
I can confirm that this solves my problem with numba + distributed. Thanks for the quick fix!

File: numba/dispatcher.py (hunk 1 above)
Comment (pitrou) [inline numba/dispatcher.py]: Not really on most mainstream systems, I guess. The `uuid` module can do some weird stuff on some systems, though (such as launch a system executable to get the MAC address).

**Comment by seibert** [comment]
@sklam: Any objections?  Will this interfere with any CUDA serialization stuff you've done recently?

**Pull request merged**